Swap to explicit "development" version on `main`

jakelishman opened the pull request:
We previously have kept `main` at the hypothetical "next" minor release, then moved it "backwards" to (say) `2.1.0rc1` from `2.1.0`, branched the new stable, and then immediately bumped `main` to `2.2.0`.

<details><summary>Description of a new branching pattern we decided not to use</summary>
This changes our version workflow so that `main`'s version is monotonically increasing, and is always explicitly a "development" version.  The new workflow for the first release candidate of a minor release is:

1. create a new branch from `main` called (e.g.) `stable/2.2`.  At this point, the version number of `main` and the new branch should both be `2.2.0.dev0`.

2. on the new stable branch, bump the version to `2.2.0rc1`.

3. on `main`, bump the version to `2.3.0.dev0`.
</details>

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary



### Details and comments


We talked about this in a dev meeting a couple of weeks ago, and I forgot to actually make the PR.

jakelishman committed:
Swap to explicit "development" version on `main` for 2.3 cycle

We previously have kept `main` at the hypothetical "next" minor release,
then moved it "backwards" to (say) `2.1.0rc1` from `2.1.0`, branched the
new stable, and then immediately bumped `main` to `2.2.0`.

This changes our version workflow so that `main`'s version is
monotonically increasing, and is always explicitly a "development"
version.  The new workflow for the first release candidate of a minor
release is:

1. create a new branch from `main` called (e.g.) `stable/2.2`.  At this
   point, the version number of `main` and the new branch should both be
   `2.2.0.dev0`.

2. on the new stable branch, bump the version to `2.2.0rc1`.

3. on `main`, bump the version to `2.3.0.dev0`.

This commit also updates `docs/conf.py` to draw its version from the
Qiskit version string (to reduce the number of places we need to
update), and updates the C test harness to have it test against the Rust
expectations so that there's fewer different formats in use.

mtreinish reviewed (commented):
Code-wise this LGTM, do you want to put anything in `MAINTAINING.md` to document this policy?

jakelishman commented:
I'd looked with the intention of doing so but there's nothing in `MAINTAINING.md` at the moment about how to branch or the versioning strategy, so I thought it'd come in with the PR that adds it.  I can do it now if you prefer.

mtreinish commented:
I'd rather add something there now just so it's "documented" what our strategy is here after this change.

mtreinish commented:
This should be updated to 2.3.0dev0 now that we've release 2.2.0rc1 and `stable/2.2` has released. With that and a small description in the maintainers guide I think this should be good to merge fast.

jakelishman commented:
I've just pushed an update now that does a bit more:
- removes the hardcoded version from `docs/conf.py` in favour of just reading it from `qiskit.__version__` - we have to be able to `import qiskit` for the docs build (because of `autodoc`) anyway, so this isn't a new requirement.
- changed the C version string expectations around pre-releases to match the Rust one, so there's fewer places to update
- added some text in the maintaining guide.

jakelishman committed:
Use reasonably sized buffers in string test

mtreinish commented on crates/cext/cbindgen.toml:
```diff
@@ -12,18 +12,19 @@ after_includes = """
+#define QISKIT_RELEASE_SERIAL 0
 
-#define QISKIT_VERSION "2.2.0rc1"
+#define QISKIT_VERSION "2.3.0-dev"
```
Can we use `env!("CARGO_PKG_VERSION"))` here?

mtreinish replied:
Ah this is in a toml file, so no

jakelishman committed:
Pivot back to tagging on main

mtreinish commented on MAINTAINING.md:
```diff
@@ -4,6 +4,51 @@ This document defines a *maintainer* as a contributor with merge privileges.
+2. `qiskit-bot` will create a `stable/2.2` branch from that commit, since that's the one you should tag.
+3. on `main`, immediately push a PR that bumps the version to `2.3.0.dev0` to open development on the 2.3 series.
+
+You will need to run `cargo update` as part of a version-bump commit to propagate the changes in `Cargo.toml` to `Cargo.lock`.
```
You can just build. Running `cargo update` will update the full dependency tree based on the `Cargo.toml`.

```suggestion
You will need to run `cargo build` as part of a version-bump commit to propagate the changes in `Cargo.toml` to `Cargo.lock`.
```

jakelishman replied:
Huh, I could have sworn that I wrote `cargo build`.

mtreinish reviewed (commented):
This LGTM, just one small comment in the docs but otherwise I think we can merge this.

jakelishman committed:
Update MAINTAINING.md

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>
